### ISSUE TITLE
assert root_frame is None when using gunicorn+gevent+flask+sleep

### ISSUE THREAD

**Issue by vasansr** [joerick/pyinstrument#214]
I have looked at issue #84 with a similar description. That seems to be resolved, but when I add a delay using sleep() in my Flask API, I again get the same assertion. (I was actually trying to make an upstream call with requests.get() ... but a simpler reproduction with sleep() is what I have reported):

OS: MacOS Catalina

gunicorn.py:
```
port = "8000"
bind = "0.0.0.0:{}".format(port)

worker_class = "gevent"
daemon = False
loglevel = 'info'
```

app.py:
```
from time import sleep
from flask import Flask, g
from flask_restful import Api, Resource
from pyinstrument import Profiler

app = Flask("myapp")

@app.route("/ping")
def ping():
    return "success"

def delay():
    sleep(1)

class Hello(Resource):
    def get(self):
        delay()
        return("Hello")

api = Api(app, prefix='/api')
api.add_resource(Hello, '/hello')

DO_INSTRUMENT = True

if DO_INSTRUMENT:
    @app.before_request
    def start_profiling():
        g.profiler = Profiler(interval=0.001, async_mode='strict')
        g.profiler.start()

    @app.after_request
    def stop_profiling(response):
        g.profiler.stop()
        print("Printing profiler output")
        print(g.profiler.output_text())
        return response
```

Commandline:
```
~/.pyenv/versions/flaskbasic/bin/gunicorn -c gunicorn.py app:app --workers=1
```

pip freeze:
```
aniso8601==9.0.1
certifi==2022.9.14
charset-normalizer==2.1.1
click==8.1.3
Flask==2.2.2
Flask-RESTful==0.3.9
gevent==21.12.0
greenlet==1.1.3
gunicorn==20.1.0
idna==3.4
importlib-metadata==4.12.0
itsdangerous==2.1.2
Jinja2==3.1.2
MarkupSafe==2.1.1
pyinstrument==4.3.0
pytz==2022.2.1
six==1.16.0
urllib3==1.26.12
Werkzeug==2.2.2
zipp==3.8.1
zope.event==4.5.0
zope.interface==5.4.0
```

If I change the sleep parameter to 0.1 or 0.2, it works:
```
0.101 full_dispatch_request  flask/app.py:1798
`- 0.101 dispatch_request  flask/app.py:1773
      [4 frames hidden]  flask, flask_restful
         0.101 dispatch_request  flask_restful/__init__.py:565
         `- 0.101 get  app.py:16
            `- 0.101 delay  app.py:12
               `- 0.101 sleep  gevent/hub.py:129
                     [2 frames hidden]  gevent
```

But if the sleep parameter is 1, this is what I get:
```
Printing profiler output
[2022-09-23 18:55:56,096] ERROR in app: Exception on /api/hello [GET]
Traceback (most recent call last):
  File "/Users/vasan/.pyenv/versions/3.8.10/envs/flaskbasic/lib/python3.8/site-packages/flask/app.py", line 2525, in wsgi_app
    response = self.full_dispatch_request()
  File "/Users/vasan/.pyenv/versions/3.8.10/envs/flaskbasic/lib/python3.8/site-packages/flask/app.py", line 1823, in full_dispatch_request
    return self.finalize_request(rv)
  File "/Users/vasan/.pyenv/versions/3.8.10/envs/flaskbasic/lib/python3.8/site-packages/flask/app.py", line 1844, in finalize_request
    response = self.process_response(response)
  File "/Users/vasan/.pyenv/versions/3.8.10/envs/flaskbasic/lib/python3.8/site-packages/flask/app.py", line 2337, in process_response
    response = self.ensure_sync(func)(response)
  File "/Users/vasan/try/flaskbasic/app.py", line 35, in stop_profiling
    print(g.profiler.output_text())
  File "/Users/vasan/.pyenv/versions/3.8.10/envs/flaskbasic/lib/python3.8/site-packages/pyinstrument/profiler.py", line 295, in output_text
    return self.output(
  File "/Users/vasan/.pyenv/versions/3.8.10/envs/flaskbasic/lib/python3.8/site-packages/pyinstrument/profiler.py", line 323, in output
    return renderer.render(session)
  File "/Users/vasan/.pyenv/versions/3.8.10/envs/flaskbasic/lib/python3.8/site-packages/pyinstrument/renderers/console.py", line 43, in render
    frame = self.preprocess(session.root_frame())
  File "/Users/vasan/.pyenv/versions/3.8.10/envs/flaskbasic/lib/python3.8/site-packages/pyinstrument/session.py", line 145, in root_frame
    assert root_frame is None, ASSERTION_MESSAGE
AssertionError: Please raise an issue at https://github.com/joerick/pyinstrument/issues and let me know how you caused this error!
```

Removing the line with worker_class = "gevent" in gunicorn.py also makes the assertion go away.

**Comment by joerick** [comment]
Could you check this on `main` please @vasansr ? Much of this code has changed recently and I -think- this assertion has been removed.

**Comment by vasansr** [comment]
I checked on main and it works nicely! I also tried with an upstream request call rather than a simple sleep delay. Thank you so much.

Would you be creating a new release for this soon? I guess you can close the issue after the release.

**Comment by dlangerm** [comment]
@joerick Unfortunately I'm in a situation where my hardware prevents me from updating to python3.7 for various reasons, do you have a workaround for those of us stuck on 3.6 (and therefore unable to upgrade to the latest library version)? Thanks!

**Comment by joerick** [comment]
> Would you be creating a new release for this soon? I guess you can close the issue after the release.

Yes! This is now fixed in 4.4.0.

> do you have a workaround for those of us stuck on 3.6

I'm afraid that's quite a bit of work, the refactor in 4.4.0 was pretty big. I don't have time to do the backport, especially now [Python 3.6 is EOL](https://endoflife.date/python). If you wanted to try to fix it yourself, the solution is that when we build the frame tree, rather than trying to find the single 'root' frame, create a 'dummy' root frame and add all frames without a parent to that. If you want to try, I'm happy to look at a PR, I created a branch to target https://github.com/joerick/pyinstrument/tree/3.4.x .
